utils/fswriters.go

```diff
@@ -0,0 +1,66 @@
+package utils
+
+import (
+	"io"
+	"os"
+)
+
+type atomicFileWriter struct {
+	f        *os.File
+	path     string
+	writeErr error
+	perm     os.FileMode
+}
+
+func (a *atomicFileWriter) write(p []byte) (n int, err error) {
+	n, err = a.f.Write(p)
+	if err != nil {
+		a.writeErr = err
+	}
+	return n, err
+}
+
+func (a *atomicFileWriter) close() (err error) {
+	defer func() {
+		if err != nil || a.writeErr != nil {
+			CleanFile(a.f)
+		}
+	}()
+	if err = a.f.Sync(); err != nil {
+		a.f.Close()
+		return err
+	}
+	if err := a.f.Close(); err != nil {
+		return err
+	}
+	if err := os.Chmod(a.f.Name(), a.perm); err != nil {
+		return err
+	}
+	if a.writeErr == nil {
+		return os.Rename(a.f.Name(), a.path)
+	}
+	return a.writeErr
+}
+
+func newAtomicFileWriter(path string, perm os.FileMode) (*atomicFileWriter, error) {
+	tmpFile, err := MkTmpFile()
+	if err != nil {
+		return nil, err
+	}
+	return &atomicFileWriter{f: tmpFile, path: path, perm: perm}, nil
+}
+
+func AtomicWriteFile(filepath string, data []byte, perm os.FileMode) error {
+	f, err := newAtomicFileWriter(filepath, perm)
+	if err != nil {
+		return err
+	}
+	n, err := f.write(data)
+	if err == nil && n < len(data) {
+		f.writeErr = io.ErrShortWrite
+	}
```

Comment:
if err == nil && n < len(data) {
		f.writeErr = io.ErrShortWrite
}
we can put this part into func f.write, couldn't we?

Reply:
yes